---
title: "RNAseq ribodepleted EDA"
author: "Marco Tello"
date: "2025-04-28"
output: github_document
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Normalization tools
library(edgeR)
# Data format and ID conversion
library(data.table)
library(tidyverse)
# Plot generation 
library(ggplot2)
library(factoextra)
library(pheatmap)
# Biological annotation
library(biomaRt)

data_path <- "~/GitHub/RNAseq_ribodepleted/output/counts"
theme_set(theme_bw())

```




# Exploratory Data Analysis

The goal is to explore if:

-   Are there any outlier samples?

-   Are the count distributions are similar across all samples?

-   Do gene expression variability aligns with experimental conditions?



## Count distribution

```{r }
# Read count table
count_df <- read.table(file.path(data_path, "gene_counts.tsv"), 
                       header = TRUE) 

# Remove STAR unmapped stats
count_df <- tibble(count_df[-c(1:4),])

# Format column names
names(count_df) <- names(count_df) %>%
  str_replace("\\.RNA\\.BS\\.030325", "") %>%
  str_replace("_L001", "")

count_df %>%
  head()

```

By looking at the density plots of the expression data it is possible to determine if any sample is a clear outlier.

```{r}
count_df %>%
  pivot_longer(!gene_id) %>%
  ggplot(aes(x=value + 1, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log10', labels = scales::comma)

```

```{r}
temp <- count_df %>%
  pivot_longer(!gene_id) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) %>%
  mutate(condition = case_when(
    condition == "C" ~ "blue",
    condition == "F" ~ "red",
    condition == "L" ~ "darkgreen",
    condition == "P" ~ "purple",
    condition == "Q" ~ "pink"))


temp %>%
  ggplot(aes(x=value + 1, color= name))+
  geom_density() +
  scale_x_continuous(trans = 'log10', labels = scales::comma) +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) 


temp %>%
  filter(value > 0) %>%
  ggplot(aes(x=value, color= name))+
  geom_density() +
  scale_x_continuous(trans = 'log10', labels = scales::comma) +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) 
```

We can observe that one sample from the P group and onse sample from the Q group are slightly shifted to the left in comparison to other samples. It is possible that these samples correspond to the two samples in a different folder from the sequencing portal. 

TODO: Check sample names from the different folder.



## Analyze RNA biotypes


Given the nature of the ribodepleted sequencing, it is possible to capture RNA molecules other than mRNA. We will first explore the composition of these biotypes in the overall dataset.

```{r}

biotype_file <- file.path(data_path, "biotypes" ,"biotype_table.tsv")

if(!file.exists(biotype_file)){
  # 1. Connect to the Ensembl mart
  ensembl <- useEnsembl(
    biomart    = "genes",
    dataset    = "hsapiens_gene_ensembl",
    version    = NULL  # latest; TODO: check what version corresponds to our GTF
  )
  
  # 2. Query for gene_biotype
  genes <- count_df$gene_id %>% unique()
  bm <- getBM(
    attributes = c("ensembl_gene_id", "gene_biotype"),
    filters    = "ensembl_gene_id",
    values     = genes,
    mart       = ensembl
  )
  
  # 3. Save output in case biomaRt doesn't cooperate in the future
  write_tsv(bm, biotype_file)
}

biotype_table <- read.table(biotype_file, header = TRUE)


```


```{r}

biotype_table %>% 
  # 1) reorder gene_biotype by descending frequency
  mutate(gene_biotype = fct_infreq(gene_biotype)) %>%  
  # 2) plot
  ggplot(aes(x = gene_biotype)) +
    geom_bar(fill = 'lightblue') +
    # 3) add labels, one per bar, vertically centered
    geom_text(
      stat = 'count', 
      aes(label = after_stat(count)), 
      position = position_stack(vjust = 0.5),
      color = 'black', size = 3
    ) +
    coord_flip() +
    labs(
      x = 'Gene biotype',
      y = 'Number of ENSEMBL IDs',
      title = 'Counts per gene biotype'
    ) 



```

Based on this graph we can identify that a large number of the entries in the expression table come from non-coding mRNAs. To facilitate future re-usability of the data, we will generate different expression tables based on the ENSEMBL biotype.

```{r}

biotypes <- c("protein_coding", "lncRNA", "snRNA", "miRNA", "snoRNA")
for(current in biotypes){
  temp <- biotype_table %>%
    filter(gene_biotype == current) %>%
    dplyr::select(ensembl_gene_id)  
  current_file <- file.path(data_path, "biotypes", paste(current, "exp.tsv", sep = "_"))
  if(!file.exists(current_file)){
    count_df %>%
      filter(gene_id %in% temp$ensembl_gene_id) %>%
      write_tsv(current_file)  
  }
}

# Save CPMs for all biotypes

biotypes <- c("protein_coding", "lncRNA", "snRNA", "miRNA", "snoRNA")
cpm_data <- count_df %>%
  dplyr::select(!gene_id) %>%
  cpm(log = TRUE, prior.count = 2)
cpm_data <- data.table(gene_id = count_df$gene_id, cpm_data)
for(current in biotypes){
  temp <- biotype_table %>%
    filter(gene_biotype == current) %>%
    dplyr::select(ensembl_gene_id)  
  current_file <- file.path(data_path, "biotypes", paste(current, "log2cpm.tsv", sep = "_"))
  if(!file.exists(current_file)){
    cpm_data %>%
      filter(gene_id %in% temp$ensembl_gene_id) %>%
      write_tsv(current_file)  
  }
}

```




# Exploratory Analysis on Protein Coding Genes

## Count distribution

```{r }
# Read count table
count_df <- read.table(file.path(data_path, "biotypes", "protein_coding_exp.tsv"), 
                       header = TRUE) 

count_df %>%
  head()

```

By looking at the density plots of the expression data it is possible to determine if any sample is a clear outlier.

```{r}
count_df %>%
  pivot_longer(!gene_id) %>%
  ggplot(aes(x=value + 1, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log10', labels = scales::comma)

```

```{r}
temp <- count_df %>%
  pivot_longer(!gene_id) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) %>%
  mutate(condition = case_when(
    condition == "C" ~ "blue",
    condition == "F" ~ "red",
    condition == "L" ~ "darkgreen",
    condition == "P" ~ "purple",
    condition == "Q" ~ "pink"))


temp %>%
  ggplot(aes(x=value + 1, color= name))+
  geom_density() +
  scale_x_continuous(trans = 'log10', labels = scales::comma) +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) 


temp %>%
  filter(value > 0) %>%
  ggplot(aes(x=value, color= name))+
  geom_density() +
  scale_x_continuous(trans = 'log10', labels = scales::comma) +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) 
```

We can observe the same trend for the P and Q samples that are shifted to the left. 



## Filter lowly expressed genes

To increase data quality we removed lowly expressed genes in two steps:

1.  Removed genes with no counts across all samples.

```{r Filter no exp genes}
# Remove columns with gene information 
exp_mat <- count_df %>%
  column_to_rownames("gene_id")
# Remove rows with all entries equal zero
exp_mat <- exp_mat[!(rowSums(exp_mat == 0) == ncol(exp_mat)),]
apply(exp_mat, MARGIN = 2, quantile)
```

Based on this quick summary, seems like the problematic samples are P2 and Q3.

2.  Removed genes with less than two reads in more than 12 samples.

```{r filter lowly exp genes}

# We use the average 1st quartile as threshold (6.64), round down to 6
# floor(mean(apply(exp_mat, MARGIN = 2, quantile, 0.25)))
# Select genes to keep
min_reads <- 6
min_samples <- 10
genes_to_keep <- apply(exp_mat >= min_reads, 
                       MARGIN = 1, sum) >= min_samples

# Final gene count matrix
exp_mat <- exp_mat[genes_to_keep,]

filtered_file <- file.path(data_path, 
                           "biotypes", 
                           paste("protein_coding_exp",
                                 "_filtered",
                                 "-S", min_samples,
                                 "-R", min_reads,
                                 ".tsv", 
                                 sep = ""))
if(!file.exists(filtered_file)){
  exp_mat %>%
    rownames_to_column("gene_id") %>%
    write_tsv(filtered_file)  
}

```

Now we repeat the previous visualizations to explore how the count distributions changed after filtering

```{r warning=FALSE}
exp_mat %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(!gene_id) %>%
  ggplot(aes(x=value + 1, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log10', labels = scales::comma)

```

```{r}
temp <- exp_mat %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(!gene_id) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) %>%
  mutate(condition = case_when(
    condition == "C" ~ "blue",
    condition == "F" ~ "red",
    condition == "L" ~ "darkgreen",
    condition == "P" ~ "purple",
    condition == "Q" ~ "pink"))


temp %>%
  ggplot(aes(x=value + 1, color= name))+
  geom_density() +
  scale_x_continuous(trans = 'log10', labels = scales::comma) +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) 

```

This step made samples distributions to look mostly normal on the log10 scale.

## CPM normalization

Before proceeding to further analysis we need to normalize the gene counts, we will use *counts per million* (CPM) to adjust for library size and transform it to log space for better visualization.

```{r cpm norm}
# CPM normalization and transform to log2
expr_log2cpm <- cpm(exp_mat, 
                    log = TRUE, 
                    prior.count = 1) %>% 
  data.frame() 

expr_log2cpm %>%
  rownames_to_column(var = "gene_id") %>%
  pivot_longer(!gene_id) %>%
  ggplot(aes(x=value, color= name))+
  geom_density() 
```

Samples normalized by library size seem to remove the artifact for samples P and Q


### Violin plots

To directly compare sample gene expression distribution without overlap between density plots we generated the respective violin plots

```{r violin plot}

temp <- expr_log2cpm %>%
  rownames_to_column(var="gene_id") %>%
  pivot_longer(!gene_id) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) 
# Convert condition to factor 
temp <- temp %>%
  mutate(condition = factor(condition, 
                            levels =c("C","F","L","P","Q")))

# Plot using ggplot with correct fill mapping
temp %>%
  ggplot(aes(x=name, y=value, fill = condition)) +
  geom_violin()+
  scale_fill_manual(values = c("C" = "blue", 
                               "F" = "red", 
                               "L" = "darkgreen", 
                               "P" = "purple",
                               "Q" = "pink")) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

```

Utilizing log2(CPM) values we can observe that all samples have similar expression distributions.

### Sample-sample correlation plot

Using normalized counts we generated a plot to observe if there is correlation within experimental groups such that samples are clustered accordingly.

```{r sample-sample}
# Annotation

# Standardize conditon names
condition <- names(expr_log2cpm) %>% 
  str_replace(pattern = "\\d_S\\d+", 
              replacement = "")

annot <- data.frame(Diet = as.factor(condition),
                    row.names = names(expr_log2cpm))

annot_colors <- list(Diet = c("C" = "blue", 
                              "F" = "red", 
                              "L" = "darkgreen", 
                              "P" = "purple",
                              "Q" = "pink"))

# Heatmap            
expr_log2cpm %>%
  cor(method = "spearman") %>%
  pheatmap(annotation_col = annot,
           annotation_row = annot,
           show_rownames = FALSE, 
           annotation_colors = annot_colors, 
           angle_col = 45)


```

### Principal Component Analysis (PCA)

Another way of determining relationship between the samples is through a PCA analysis, which reduces the dimentionality of our data to a set of independent variables (principal components) that represent the major proportion of variability in our data.

```{r PCA CPM}

PCs <- prcomp(t(cpm(exp_mat)), center = TRUE, scale = TRUE)
# Scree plot 
fviz_eig(PCs)
# Scatter plot
eig_val <- get_eigenvalue(PCs)
PCs <- cbind(annot, PCs$x[,1:10])
PCs$sample_id <- rownames(PCs)

PCs <- PCs %>%
  mutate(color_class = case_when(
    condition == "C" ~ "blue",
    condition == "F" ~ "red",
    condition == "L" ~ "darkgreen",
    condition == "P" ~ "purple",
    condition == "Q" ~ "orange")) 

PCs %>%
  ggplot(aes(x = PC1, y = PC2, 
             color = condition)) + 
  geom_point(aes(size = 8)) + 
  scale_color_manual(values = PCs$color_class, 
                     breaks = PCs$Diet) +
  labs(x= paste("PC1 (",round(eig_val$variance.percent[1], 2),"%)", sep = ""),
       y= paste("PC2 (",round(eig_val$variance.percent[2], 2),"%)", sep = ""))+
  guides(size = "none") +
  theme_bw()


PCs %>%
  ggplot(aes(x = PC3, y = PC2, 
             color = Diet)) + 
  geom_point(aes(size = 8)) + 
  scale_color_manual(values = PCs$color_class, 
                     breaks = PCs$Diet) +
  labs(x= paste("PC3 (",round(eig_val$variance.percent[3], 2),"%)", sep = ""),
       y= paste("PC2 (",round(eig_val$variance.percent[2], 2),"%)", sep = ""))+
  guides(size = "none") +
  theme_bw()


```

# Exploratory Analysis on lncRNA

## Count distribution

```{r}
# Read count table
count_df <- read.table(file.path(data_path, "biotypes", "lncRNA_log2cpm.tsv"), 
                       header = TRUE) 

count_df %>%
  head()

```

By looking at the density plots of the expression data it is possible to determine if any sample is a clear outlier.

```{r}
count_df %>%
  pivot_longer(!gene_id) %>%
  ggplot(aes(x=value, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log2', labels = scales::comma)

```

```{r}
temp <- count_df %>%
  pivot_longer(!gene_id) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) %>%
  mutate(condition = case_when(
    condition == "C" ~ "blue",
    condition == "F" ~ "red",
    condition == "L" ~ "darkgreen",
    condition == "P" ~ "purple",
    condition == "Q" ~ "pink"))


temp %>%
  ggplot(aes(x=value, color= name))+
  geom_density() +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name)+
  scale_x_continuous(trans = 'log2', labels = scales::comma) 


temp %>%
  filter(value > 0) %>%
  ggplot(aes(x=value, color= name))+
  geom_density() +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) +
  scale_x_continuous(trans = 'log2', labels = scales::comma)

```

We can observe the same trend for the P and Q samples that are shifted to the left. 



## Filter lowly expressed lncRNA

To increase data quality we removed lowly expressed lncRNA in two steps:

1.  Removed lncRNA with no counts across all samples.

```{r}
# Remove columns with gene information 
exp_mat <- count_df %>%
  column_to_rownames("gene_id")
# Remove rows with all entries equal zero
exp_mat <- exp_mat[!(rowSums(exp_mat == 0) == ncol(exp_mat)),]
```

Now we repeat the previous visualizations to explore how the count distributions changed after filtering

```{r warning=FALSE}
exp_mat %>%
  rownames_to_column("lncRNA") %>%
  pivot_longer(!lncRNA) %>%
  ggplot(aes(x=value, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log2', labels = scales::comma)

```

This step made samples distributions to look mostly normal on the log2 scale.

## Violin plots

To directly compare sample gene expression distribution without overlap between density plots we generated the respective violin plots

```{r }

temp <- exp_mat %>%
  rownames_to_column(var="lncRNA") %>%
  pivot_longer(!lncRNA) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) 
# Convert condition to factor 
temp <- temp %>%
  mutate(condition = factor(condition, 
                            levels =c("C","F","L","P","Q")))

# Plot using ggplot with correct fill mapping
temp %>%
  ggplot(aes(x=name, y=value + 4, fill = condition)) +
  geom_violin()+
  scale_fill_manual(values = c("C" = "blue", 
                               "F" = "red", 
                               "L" = "darkgreen", 
                               "P" = "purple",
                               "Q" = "pink")) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))+
  scale_y_continuous(trans = 'log2', labels = scales::comma)

```


## Sample-sample correlation plot

Using normalized counts we generated a plot to observe if there is correlation within experimental groups such that samples are clustered accordingly.

```{r }
# Annotation

# Standardize conditon names
condition <- names(exp_mat) %>% 
  str_replace(pattern = "\\d_S\\d+", 
              replacement = "")

annot <- data.frame(Diet = as.factor(condition),
                    row.names = names(exp_mat))

annot_colors <- list(Diet = c("C" = "blue", 
                              "F" = "red", 
                              "L" = "darkgreen", 
                              "P" = "purple",
                              "Q" = "pink"))

# Heatmap            
exp_mat %>%
  cor(method = "spearman") %>%
  pheatmap(annotation_col = annot,
           annotation_row = annot,
           show_rownames = FALSE, 
           annotation_colors = annot_colors, 
           angle_col = 45)


```


# Exploratory Analysis on miRNA

## Count distribution

```{r}
# Read count table
count_df <- read.table(file.path(data_path, "biotypes", "miRNA_log2cpm.tsv"), 
                       header = TRUE) 

count_df %>%
  head()

```

By looking at the density plots of the expression data it is possible to determine if any sample is a clear outlier.

```{r}
count_df %>%
  pivot_longer(!gene_id) %>%
  ggplot(aes(x=value+4, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log2', labels = scales::comma)

```

```{r}
temp <- count_df %>%
  pivot_longer(!gene_id) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) %>%
  mutate(condition = case_when(
    condition == "C" ~ "blue",
    condition == "F" ~ "red",
    condition == "L" ~ "darkgreen",
    condition == "P" ~ "purple",
    condition == "Q" ~ "pink"))


temp %>%
  ggplot(aes(x=value+4, color= name))+
  geom_density() +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name)+
  scale_x_continuous(trans = 'log2', labels = scales::comma) 


temp %>%
  filter(value > 0) %>%
  ggplot(aes(x=value, color= name))+
  geom_density() +
  scale_color_manual(values = temp$condition, 
                     breaks = temp$name) +
  scale_x_continuous(trans = 'log2', labels = scales::comma)

```

We can observe the same trend for the P and Q samples that are shifted to the left. 



## Filter lowly expressed lncRNA

To increase data quality we removed lowly expressed lncRNA in two steps:

1.  Removed lncRNA with no counts across all samples.

```{r}
# Remove columns with gene information 
exp_mat <- count_df %>%
  column_to_rownames("gene_id")
# Remove rows with all entries equal zero
exp_mat <- exp_mat[!(rowSums(exp_mat == 0) == ncol(exp_mat)),]
```

Now we repeat the previous visualizations to explore how the count distributions changed after filtering

```{r warning=FALSE}
exp_mat %>%
  rownames_to_column("miRNA") %>%
  pivot_longer(!miRNA) %>%
  ggplot(aes(x=value, color = name)) +
    geom_density() +
    scale_x_continuous(trans = 'log2', labels = scales::comma)

```

This step made samples distributions to look mostly normal on the log2 scale.


## Violin plots

To directly compare sample gene expression distribution without overlap between density plots we generated the respective violin plots

```{r }

temp <- exp_mat %>%
  rownames_to_column(var="miRNA") %>%
  pivot_longer(!miRNA) %>%
  mutate(condition = str_replace(string = name, 
                                 pattern = "\\d_S\\d+", 
                                 replacement = "")) 
# Convert condition to factor 
temp <- temp %>%
  mutate(condition = factor(condition, 
                            levels =c("C","F","L","P","Q")))

# Plot using ggplot with correct fill mapping
temp %>%
  ggplot(aes(x=name, y=value + 4, fill = condition)) +
  geom_violin()+
  scale_fill_manual(values = c("C" = "blue", 
                               "F" = "red", 
                               "L" = "darkgreen", 
                               "P" = "purple",
                               "Q" = "pink")) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))+
  scale_y_continuous(trans = 'log2', labels = scales::comma)

```


### Sample-sample correlation plot

Using normalized counts we generated a plot to observe if there is correlation within experimental groups such that samples are clustered accordingly.

```{r }
# Annotation

# Standardize conditon names
condition <- names(exp_mat) %>% 
  str_replace(pattern = "\\d_S\\d+", 
              replacement = "")

annot <- data.frame(Diet = as.factor(condition),
                    row.names = names(exp_mat))

annot_colors <- list(Diet = c("C" = "blue", 
                              "F" = "red", 
                              "L" = "darkgreen", 
                              "P" = "purple",
                              "Q" = "pink"))

# Heatmap            
exp_mat %>%
  cor(method = "spearman") %>%
  pheatmap(annotation_col = annot,
           annotation_row = annot,
           show_rownames = FALSE, 
           annotation_colors = annot_colors, 
           angle_col = 45)


```